matrix/square_matrix.py

```diff
@@ -91,11 +89,97 @@ def LU_decomposition(self):
+            col_num: <int> number of columns in the Toeplitz matrix, equal to n
+        Returns:
+            L: <Matrix> (n + 1) x n Toeplitz lower triangular matrix
+
+        For more info, see:
+        https://handwiki.org/wiki/Samuelson%E2%80%93Berkowitz_algorithm#:~:text=In%20mathematics%2C%20the%20Samuelson%E2%80%93Berkowitz,commutative%20ring%20without%20zero%20divisors.
+        """
+        row_num = col_num + 1  # number of rows of Toeplitz matrix
+        L = Matrix(self.identity(col_num).ls_entries + [[0] * (col_num)])
+        products = [-a_0_0, -row_vector.__mul__(col_vector.transpose())[0][0]]
+
+        # get the diagonal entries where taking the power of the principal submatrix is required
+        if row_num > 3:
+            principal_pows = [principal] + [principal.__pow__(x) for x in range(2, col_num)]
+            products += [-row_vector.__mul__(x).__mul__(col_vector.transpose())[0][0] for x in principal_pows]
+
+        # Assign values to the diagonals
+        for i in range(1, row_num):
+            for j in range(col_num):
+                if i - j > 0:
+                    L[i][j] = products[(i - j) - 1]
+        return L
+
+    def char_eqn_berkowitz(self):
+        """
+        Gets the coefficients of the characteristic polynomial using the Berkowitz Algorithm
+        # https://en.wikipedia.org/wiki/Samuelson%E2%80%93Berkowitz_algorithm
+
+
+        Returns:
+            <Matrix> vector containing the coefficients of the characteristic polynomial from highest to lowest degree
+        """
+        A = SquareMatrix(copy.deepcopy(self.ls_entries))
+        C_ls = []
+
+        for i in range(self.size - 1):
+            # partition the n x n matrix accordingly
+            row_vector = Matrix([A[0, 1:]])
+            col_vector = Matrix([A[1:, 0]])
+            principal = Matrix([[A[j][x] for x in range(len(A[j])) if x != 0] for j in range(len(A.ls_entries)) if
+                                j != 0])
+            a_elem = A[0][0]
+            C = self.get_toeplitz_matrix_berkowitz(a_elem, row_vector, col_vector, principal, col_num=len(A.ls_entries))
+            C_ls.append(C)
+            A = principal  # next Toeplitz matrix is found for subsequent principal submatrix
+
+        # last one is just [1,-a,1,1]
+        C_ls.append(Matrix(ls_entries=[[1, -A[0][0]]]).transpose())
+
+        # create the resulting vector that stores the coefficients of the characteristic polynomial
+        result = C_ls[0]
+        C_ls.pop(0)
+        while C_ls:
+            result = result.__mul__(C_ls[0])
+            C_ls.pop(0)
+
+        return result.transpose()
+
+    def eigenvalues(self):
+        """
+        Uses the numpy roots function to calculate the roots of the characteristic polynomial
+        Characteristic equation is found using the Berkowitz Algorithm
+        """
+        char_eqn = self.char_eqn_berkowitz()
+        return list(numpy_roots(char_eqn.ls_entries[0]))  # this uses the numpy roots function
```

Comment:
This part is using the numpy roots function for root finding... not sure if we want to implement another root finding function

Reply:
haha yeah, probably we want, but we can do that later. I had a look at how the numpy roots functions works it actually computes the eigenvalues of another matrix to compute these roots haha.

Reply:
Haha sounds good!